src/main/java/net/minecraftforge/resource/ResourceCacheManager.java

```diff
@@ -69,54 +73,81 @@
+    public ResourceCacheManager(final boolean supportsReloading, final ForgeConfigSpec.BooleanValue indexOffThreadConfig, final BiFunction<PackType, String, Path> pathBuilder) {
         this.supportsReloading = supportsReloading;
-        this.indexOffThreadConfigOption = indexOffThreadConfig;
+        this.indexOnThreadConfigurationKey = Iterators.getLast(indexOffThreadConfig.getPath().iterator());
         this.pathBuilder = pathBuilder;
     }
 
-    public boolean shouldIndexOffThread()
-    {
-        return !getConfigValue(this.indexOffThreadConfigOption);
+    public ResourceCacheManager(final boolean supportsReloading, final String indexOffThreadConfig, final BiFunction<PackType, String, Path> pathBuilder) {
+        this.supportsReloading = supportsReloading;
+        this.indexOnThreadConfigurationKey = indexOffThreadConfig;
+        this.pathBuilder = pathBuilder;
     }
 
-    public static boolean shouldUseCache()
-    {
-        return getConfigValue(ForgeConfig.COMMON.cachePackAccess);
+    public static boolean shouldUseCache() {
+        return getConfigValue("cachePackAccess", true);
     }
 
-    private static boolean getConfigValue(Supplier<Boolean> configValue)
-    {
-        // Yes we catch the early loading error on purpose.
-        // Both the indexing off-thread and if caching is enabled are configurable, so we need to catch the error.
-        // So when they are loaded early we preserve the default behaviour which is to not cache or index on thread.
-        try
-        {
-            return configValue.get();
-        } catch (IllegalStateException e)
-        {
-            return false;
+    private static boolean getConfigValue(final String configKey, final boolean defaultValue) {
+        final Path configPath = FMLPaths.CONFIGDIR.get().resolve("forge-resource-caching.toml");
+        if (!Files.exists(configPath)) {
+            try {
+                Files.write(configPath, ImmutableList.of(
+                                "# This TOML configuration file controls the resource caching system which is used before the mod loading environment starts.",
+                                "# This file is read by the Forge boot loader, and is not used by the game itself.",
+                                "#",
+                                "# Set this to false to disable the resource cache. This will cause the game to scan the resource packs everytime it needs a list of resources.",
+                                "cacheResources=true",
+                                "",
+                                "# Set this to false to forge the caching of vanilla resources to happen on the main thread.",
+                                "indexVanillaPackCachesOnThread=false",
+                                "",
+                                "# Set this to false to forge the caching of mod resources to happen on the main thread.",
+                                "indexModPackCachesOnThread=false"
```

Comment:
This was what was requested by lex, not sure if it is such a good idea to hardcode.

The basic idea is to hide the fact that forge needs configs (and loads them) before mod loading as much as possible from modders, so that they do not try to load their own configs before registries are loaded.

I can make it a file, I don't care.

Reply:
I will leave it a hardcode for now, should prevent people from being nut jobs and messing with it.

Reply:
Could you please elaborate on what you mean by "people [...] messing with it"?